tests/orchestrations_test.go

```diff
@@ -715,6 +716,173 @@ func Test_TerminateOrchestration(t *testing.T) {
+
+			// Terminate the root orchestration and mark whether a recursive termination
+			output := fmt.Sprintf("Recursive termination = %v", recurse)
+			opts := []api.TerminateOptions{api.WithOutput(output), api.WithRecursiveTerminate(recurse)}
+			require.NoError(t, client.TerminateOrchestration(ctx, id, opts...))
+
+			// Wait for the root orchestration to complete and verify its terminated status
+			metadata, err := client.WaitForOrchestrationCompletion(ctx, id)
+			require.NoError(t, err)
+			require.Equal(t, protos.OrchestrationStatus_ORCHESTRATION_STATUS_TERMINATED, metadata.RuntimeStatus)
+			require.Equal(t, fmt.Sprintf("\"%s\"", output), metadata.SerializedOutput)
+
+			// Wait for all L2 suborchestrations to complete
+			orchIDs := []string{}
+			for i := 0; i < 5; i++ {
+				orchIDs = append(orchIDs, string(id)+"_L1_"+strconv.Itoa(i)+"_L2")
+			}
+			for _, orchID := range orchIDs {
+				_, err := client.WaitForOrchestrationCompletion(ctx, api.InstanceID(orchID))
+				require.NoError(t, err)
+			}
+			// Verify tht none of the L2 suborchestrations executed the activity in case of recursive termination
+			assert.NotEqual(t, recurse, executedActivity)
+		})
+	}
+}
+
+func Test_TerminateOrchestration_Recursive_TerminateCompletedSubOrchestration(t *testing.T) {
+	delayTime := 4 * time.Second
+	r := task.NewTaskRegistry()
+	r.AddOrchestratorN("Root", func(ctx *task.OrchestrationContext) (any, error) {
+		ctx.CallSubOrchestrator("L1", task.WithSubOrchestrationInstanceID(string(ctx.ID)+"_L1")).Await(nil)
+		ctx.CreateTimer(delayTime).Await(nil)
+		return nil, nil
+	})
+	r.AddOrchestratorN("L1", func(ctx *task.OrchestrationContext) (any, error) {
+		// Not waiting for L2 to complete
+		ctx.CallSubOrchestrator("L2", task.WithSubOrchestrationInstanceID(string(ctx.ID)+"_L2"))
+		return nil, nil
+	})
+	r.AddOrchestratorN("L2", func(ctx *task.OrchestrationContext) (any, error) {
+		ctx.CreateTimer(delayTime).Await(nil)
+		return nil, nil
+	})
+
+	// Initialization
+	ctx := context.Background()
+	client, worker := initTaskHubWorker(ctx, r)
+	defer worker.Shutdown(ctx)
+
+	// Test terminating with and without recursion
+	for _, recurse := range []bool{true, false} {
+		t.Run(fmt.Sprintf("Recurse = %v", recurse), func(t *testing.T) {
+			// Run the orchestration, which will block waiting for external events
+			id, err := client.ScheduleNewOrchestration(ctx, "Root")
+			require.NoError(t, err)
+
+			// Wait long enough to ensure that all L1 orchestrations have completed but Root and L2 are still running
+			assert.Eventually(t, func() bool {
+				orchestrationIDs := []string{string(id), string(id) + "_L1", string(id) + "_L1_L2"}
+				for _, orchID := range orchestrationIDs {
+					metadata, err := client.FetchOrchestrationMetadata(ctx, api.InstanceID(orchID))
+					require.NoError(t, err)
+					if orchID == string(id)+"_L1" {
+						if metadata.RuntimeStatus != protos.OrchestrationStatus_ORCHESTRATION_STATUS_COMPLETED {
```

Comment:
Added some more comments. 

The statusChecks here are just to ensure that before sending the terminate event, L1 orchestration should have completed but Root and L2 orchestration should be running, so that we check for the testcase where we terminate an already completed sub-orchestration whose child sub-orchestration is still running